src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -581,10 +581,17 @@ private void OnImportChanged(object sender, EventArgs e)
 
         private void OnAnalyzerChanged(object sender, EventArgs e)
         {
-            VisualStudioAnalyzer analyzer = (VisualStudioAnalyzer)sender;
-
-            RemoveAnalyzerAssembly(analyzer.FullPath);
-            AddAnalyzerAssembly(analyzer.FullPath);
+            // Postpone handler's actions to prevent deadlock. This AnalyzeChanged event can
+            // be invoked while the FileChangeService lock is held, and VisualStudioAnalyzer's 
+            // efforts to listen to file changes can lead to a deadlock situation.
+            // Postponing the VisualStudioAnalyzer operations gives this thread the opportunity
+            // to release the lock.
+            Dispatcher.CurrentDispatcher.BeginInvoke(new Action(() => {
```

Comment:
Yes, this was raised on the UI thread in the stack in #11019. I suppose it may be invoked in a different thread in other scenarios. But doesn't `Dispatcher.CurrentDispatcher` invoke the operation in the current thread? 

I did not want to change the semantics of this method such that its operations could be performed in a different thread.